### ISSUE TITLE
receiver: socket daemon like OpenCensus PHP has

### ISSUE THREAD

**Issue by odeke-em** [census-instrumentation/opencensus-service#389]
Recording this here for posterity after the OpenCensus community meeting of February 13th 2019, I briefly talked about OpenCensus PHP and the little Go daemon that receives traffic via a socket.  The utility raised is that the daemon can perform aggregation then finally upload to OpenCensus-Agent.

Turns out that OpenCensus-Web will need the same functionality but also it might be useful to decouple the daemon in https://github.com/census-instrumentation/opencensus-php/tree/d1512abf456761165419a7b236e046a38b61219e/daemon
bring it to the OpenCensus Agent and that way OpenCensus PHP can work seamlessly and without having to run the daemon, as well as cut short the network and bandwidth costs by directly dumping traces and metrics to the OpenCensus Agent sink since they'll share the same process.

Thanks to @bogdandrutu for reminding me. Kindly paging @draffensperger @basvanbeek

**Comment by draffensperger** [comment]
Thanks for bringing this up. For the OpenCensus Web use case, ideally there would be a way to specify the metrics and views in a config file that the customer uploads with their OC agent. Then the web library just writes the specific measurements with labels/values to the named metrics.

What I had initially envisioned was having the OC agent implement the [stats](https://github.com/census-instrumentation/opencensus-proto/blob/master/src/opencensus/proto/stats/v1/stats.proto) OpenCensus proto. I would think for the OC Web use case, the `Measure` and `View` data would be in a config file, but there would be an HTTP-callable endpoint to write raw `Measurement` values.

However, if doing this via Prometheus is better, I'm open to that. I'll plan to look at the PHP example more to see if there would be a natural way to adapt that to the web case of configuring the metrics in a file on the server and then writing measurements via HTTP from the web clients.

**Comment by basvanbeek** [comment]
In PHP this is how it works:
- you run a PHP stats initialize script at service start time which will create measurements & views and registers them with the Daemon.
- in request handlers you will have to recreate the measurements you need for handling that specific call, using the same names as found in the in the initialize script and then you can record them.

This will have the Daemon register the measurements and views once and then just wait on incoming raw stats recordings when PHP is serving requests. 

If not doing an initialize script and reregistering measurements and views constantly for each request, it will still work but you incur additional transport and cpu cost.

@draffensperger please take a look at the Daemon protocol documentation as found in the link sent by @odeke-em and see if it would fit your needs. For quick testing you can run the PHP ecosystem from the following docker-compose project: https://github.com/basvanbeek/opencensus-php-docker

The unix socket of the Daemon to write to can be found in a shared volume which is mounted under `./data/daemon/socket/oc-daemon.sock`

**Comment by flands** [comment]
Behavior of this will change in the OpenTelemetry Service given the merge of Agent and Collector into a single binary. If this is still an issue, please open a new issue in the OpenTelemetry Service repo.
